### ISSUE TITLE
How to clean the TGI guidance cache?

### ISSUE THREAD

**Issue by EdisonE3** [huggingface/text-generation-inference#2200]
I use TGI guidance to enforce LLM choose a tool.
However, when I change the description of the tool, I find TGI does not re-compile the new grammar.
Therefore, I want to know how to clean the compiled grammar.

**Comment by ErikKaum** [comment]
Thanks for reporting this @EdisonE3!

Was there any particular model/setup that this happened on or does it (which I assume) happen globally?

@drbh cool if I look into this? I might ping you if I get stuck but I'd suspect the _lru_cache isn't getting purged?

**Comment by EdisonE3** [comment]
Thanks for your response!

**Comment by EdisonE3** [comment]
@ErikKaum Thanks for your response! 
I use Qwen2-7B-Instruct and launch it through: `text-generation-launcher --model-id Qwen2-7B-Instruct --port 8000 --max-total-tokens 8000 --max-input-tokens 6000 --num-shard 2`.
I think field description modifying not trigger a re-compile may be reasonable since it do not affect the grammar structure.

Additionally, if it's not too much trouble, could you please let me know which fields in the tool description will trigger grammar compilation or where I can find them?

**Comment by ErikKaum** [comment]
Sorry for a slow response!

I was able to replicate locally the behavior you described 👍 But I'm not able to confirm if it's that the new grammar doesn't get compiled vs. limitations in the model itself. I noticed that `Qwen2-7B-Instruct` requires quite a lot of "guiding" to get things right.

So about here is the entrypoint of the `v1/chat/completions` where then `tools_grammar_prompt` is baked into the chat template in the `let inputs = match infer.apply_chat_template...` function

https://github.com/huggingface/text-generation-inference/blob/da82c63a4ff9c6b8f3d0901cb955c8db04c9a492/router/src/server.rs#L1070-L1092

There's also a corresponding place in the python code which I haven't double checked:
https://github.com/huggingface/text-generation-inference/blob/da82c63a4ff9c6b8f3d0901cb955c8db04c9a492/server/text_generation_server/utils/logits_process.py#L483-L529

Hopefully these help 👍
